### PR TITLE
Add basic segment test for RAJA Teams

### PR DESCRIPTION
**Pull request by artv3**
Building up the testing framework for RAJA Teams. This PR test Range/Stride RangeSegments and Listsegments with RAJA loop methods. 

Note this PR reworks listsegment so they are device friendly when captured via lambda.

### CODE REVIEW

**Comment by artv3** [comment]
Had to switch to drafts, just noticed some of the test failed.

**Comment by rhornung67** [comment]
@artv3 note that lassen is down today for upgrades and we ROCm 4.5.2 was removed from corona, so we have to update to 5.x to make testing run there.

**Comment by artv3** [comment]
ping: @rhornung67 @rchen20 @mdavis36

File: test/include/RAJA_test-teams-runtime-execpol.hpp
Comment (rchen20) [inline test/include/RAJA_test-teams-runtime-execpol.hpp:101]: Are policies like these specifically for the Teams host & device selection mechanism?

**Review by rchen20 — DISMISSED** [review]

**Comment by artv3** [comment]
Did anyone else want to take a look? or can I merge this PR? @rhornung67  @mdavis36 ?

File: test/functional/teams/segment/tests/test-teams-ListSegment.hpp
Comment (rhornung67) [inline test/functional/teams/segment/tests/test-teams-ListSegment.hpp:81]: At this point, we're inside a GPU kernel potentially and we are using the list segment in the same way we would pass it from a non-GPU context. The segment indices should be in properly device memory. Do we need to specialize the RAJA::expt::loop construct for list segments?

Comment (artv3) [reply test/functional/teams/segment/tests/test-teams-ListSegment.hpp:81]: I believe the issue is that we need to capture the list segment into the host_device lambda which triggers the copy and the destructor of the listsegment object. 

https://github.com/LLNL/RAJA/blob/87a5cac67214e5e96c941bd652b1c0981e9f2123/include/RAJA/index/ListSegment.hpp#L174-L180

https://github.com/LLNL/RAJA/blob/87a5cac67214e5e96c941bd652b1c0981e9f2123/include/RAJA/index/ListSegment.hpp#L192-L197

I think what we want is to take a reference of the device pointer and use that within the host_device lambda. 
Perhaps wrapping it into a struct could work?

```
struct list_seg_ptr
{
double * device_ptr; //point to listsegment pointer
int begin;
int end;
};
```

I think we would have the same issue with we tried to use a listsegment inside a raja forall method..
```
RAJA::listSegment lseg;
  RAJA::forall<RAJA::cuda_exec<CUDA_BLOCK_SIZE>>(RAJA::RangeSegment(0, N), 
    [=] RAJA_DEVICE (int i) { 
     lseg.size();  
    c[i] = a[i] + b[i]; 
  });
```    

Let me try the above ^^.

**Comment by rhornung67** [comment]
Let's get a few more folks to look over and approve before merging.  Please add an issue about making list segment work with teams. This should be a discussion topic at the next RAJA meeting.

**Comment by artv3** [comment]
> Let's get a few more folks to look over and approve before merging. Please add an issue about making list segment work with teams. This should be a discussion topic at the next RAJA meeting.

Sure sounds good, I can make a small example to walk through the meeting.

File: test/functional/teams/segment/tests/test-teams-ListSegment.hpp
Comment (MrBurmark) [reply test/functional/teams/segment/tests/test-teams-ListSegment.hpp:81]: Does the `m_owned` tracking not take care of this?

Comment (artv3) [reply test/functional/teams/segment/tests/test-teams-ListSegment.hpp:81]: I made a small example in PR https://github.com/LLNL/RAJA/pull/1266 , if I use the listsegment directly inside the kernel then I hit this error:
 
```
terminate called after throwing an instance of 'std::runtime_error'
  what():  campCudaErrchk(cudaStreamSynchronize(stream)) unspecified launch failure /g/g17/vargas45/RAJA_DIR/RAJA/tpl/camp/include/camp/resource/cuda.hpp:172
Aborted
```

but if I use the proof concept wrapper it seems to work. @MrBurmark that is a good point about the data not being owned by the object copy, seems like we shouldn't be triggering the free.

**Comment by artv3** [comment]
@rhornung67 we think the issue could be on the copy of m_resource for listsegment: 
https://github.com/LLNL/RAJA/blob/5f290e72c64fff60c6c9d206de8a730106102ce3/include/RAJA/index/ListSegment.hpp#L175

@MrBurmark mentioned it has a std::shared_ptr so there is some activity on the CPU, which is likely causing the segfault on the device. 
Does sound right @trws @mdavis36  ?

**Comment by trws** [comment]
That is very possible, Will have to give some careful thought to how we can avoid that problem.  One unfortunate but expedient way would be to give resources a way to clone themselves generically, then hold onto a heap-allocated copy of the type-erased resource by pointer.  That's really where this nastiness is coming from, it's an erased type we can't make statically typed without breaking a whole lot of stuff, so we probably either need a specialized smart pointer to avoid the device copy or to do something nasty like I outlined above.

**Comment by trws** [comment]
Wrapping the iterator would probably be a good solution if that's an option, the trick with wrapping the resource is ensuring that it doesn't get destroyed early, it may not come up but that's the concern.

**Comment by artv3** [comment]
> Wrapping the iterator would probably be a good solution if that's an option, the trick with wrapping the resource is ensuring that it doesn't get destroyed early, it may not come up but that's the concern.

I went ahead and pushed a potential solution where we wrap the iterator within RAJA::Span. If that is the direction we would like to go we can merge it in with this PR; otherwise I can revert the change.

**Comment by trws** [comment]
Since the iterator doesn't carry the resource, that's a really simple solution to the problem.  My one gripe is I'd prefer an additional single-argument overload to `make_span` [here](https://github.com/LLNL/RAJA/blob/9e12cc2a1460b0523ff100ceae9b98a2446ad41a/include/RAJA/util/Span.hpp#L160) rather than a method on ListSegment.  Not a huge difference, but using the usual `using begin; begin(iterable)` etc. dance to grab the iterators should work for all iterables and save us from redoing it on others later.

**Comment by artv3** [comment]
> Since the iterator doesn't carry the resource, that's a really simple solution to the problem. My one gripe is I'd prefer an additional single-argument overload to `make_span` [here](https://github.com/LLNL/RAJA/blob/9e12cc2a1460b0523ff100ceae9b98a2446ad41a/include/RAJA/util/Span.hpp#L160) rather than a method on ListSegment. Not a huge difference, but using the usual `using begin; begin(iterable)` etc. dance to grab the iterators should work for all iterables and save us from redoing it on others later.

Thanks, just pushed it up.

File: include/RAJA/util/Span.hpp
Comment (trws) [inline include/RAJA/util/Span.hpp:169]: Since we're C++14 now, the trailing return type is optional.  Not a problem to have it there since this is what comes out anyway, but we no longer need to repeat them like this.

**Comment by artv3** [comment]
@MrBurmark  @rhornung67 I pushed up the code that considers m_resource as a pointer in listsegment. The following test will fail: test-forall-IndexSet-Sequential.exe . 
However; teams_listsegments will now work.

File: include/RAJA/index/ListSegment.hpp
Comment (MrBurmark) [inline include/RAJA/index/ListSegment.hpp:125]: initIndexData should probably handle allocating m_resource, as it should not be done if Unowned.

File: include/RAJA/index/ListSegment.hpp
Comment (MrBurmark) [inline include/RAJA/index/ListSegment.hpp:146]: m_resource should be nullptr since Unowned.

File: include/RAJA/index/ListSegment.hpp
Comment (MrBurmark) [inline include/RAJA/index/ListSegment.hpp:182]: Already done during initialization.

File: include/RAJA/index/ListSegment.hpp
Comment (MrBurmark) [inline include/RAJA/index/ListSegment.hpp:196]: Need to set rhs.m_resource to nullptr.

File: include/RAJA/index/ListSegment.hpp
Comment (MrBurmark) [inline include/RAJA/index/ListSegment.hpp:205]: Move into a new method like `clear()` so it can be reused in copy assignment and move assignment operators.

File: include/RAJA/index/ListSegment.hpp
Comment (MrBurmark) [inline include/RAJA/index/ListSegment.hpp:198]: Need to handle copy assignment and move assignment.

Comment (artv3) [reply include/RAJA/index/ListSegment.hpp:197]: Thanks for the feedback! just pushed some changes, which copy and move assignments do you mean here?

File: examples/CMakeLists.txt
Comment (artv3) [inline examples/CMakeLists.txt:37]: revert chages

File: examples/teams_listsegments.cpp
Comment (artv3) [inline examples/teams_listsegments.cpp:8]: remove

**Comment by artv3** [comment]
> @artv3 @MrBurmark I pushed some changes to fix the broken test and add more checks to it to make it more robust. If all checks pass, let's debate whether we want to change the ownership semantics (e.g., make all copy operations deep copies and remove the unowned concept).

I think if we we want to use ListSegment in some threaded environement we could consider remove the resource component of it, and have the user manage memory. Essentially have the same behavior as span. 

If we do want to continue managing memory for ListSegment maybe we can change the name to ListContainer and make a note that containers are assumed to have ownership of the data. Users must use a reference to the data inside a RAJA kernel.

**Comment by rhornung67** [comment]
@artv3 should we merge this first and then do the list segment changes in a separate PR? I can start on that.

**Comment by artv3** [comment]
> @artv3 should we merge this first and then do the list segment changes in a separate PR? I can start on that.

Yes, sounds good!

**Comment by artv3** [comment]
If I can get some approvals I can merge this in today.

**Review by rhornung67 — APPROVED** [review]

File: include/RAJA/util/Span.hpp
Comment (MrBurmark) [inline include/RAJA/util/Span.hpp:171]:
```suggestion
  using std::begin;
  using std::end;
  using std::distance;
  return Span<typename Iter::iterator, decltype(distance(begin(iterable), end(iterable)))>
    (begin(iterable), end(iterable));
```

File: examples/teams_listsegments.cpp
Comment (MrBurmark) [reply examples/teams_listsegments.cpp:8]: Should this be removed?

**Review by rhornung67 — APPROVED** [review]

File: examples/CMakeLists.txt
Comment (MrBurmark) [reply examples/CMakeLists.txt:37]: Is this reverted?

**Review by MrBurmark — APPROVED** [review]

File: examples/CMakeLists.txt
Comment (artv3) [reply examples/CMakeLists.txt]: Awk! I forgot about these! Just removed

File: examples/teams_listsegments.cpp
Comment (artv3) [reply examples/teams_listsegments.cpp]: yes! sorry good catch!

**Comment by artv3** [comment]
Should be ready now, sorry had to remove some testing files.

**Review by MrBurmark — APPROVED** [review]

**Review by rhornung67 — APPROVED** [review]

**Pull request merged**
